hyschive/gamer-fork
# Update the ELBDM test problems for the hybrid scheme

## Background
- Some ELBDM test problems do not support the hybrid scheme.
   - The phase field is wrapped (using the atan() function from the real and imaginary parts) in the initial conditions.
   - The phase needs to be unwrapped for the hybrid scheme.
- Local phase unwrapping in the solver is not what we want because the fluid scheme allows for large phase differences in adjacent cells when the velocity is large.
- Global phase unwrapping in the initial condition is enough. After initialization, we don't need the phase unwrapping during the evolution.
- The phase unwrapping cannot be done for initial conditions containing vortices. The phase discontinuity is inevitable.
- The boundary condition cannot be periodic for initial conditions with an overall velocity.
   - There will be a phase difference $\delta S=\frac{m}{\hbar}\int_0^{L_x}v_xdx$ between the boundaries.
   - We need to use the analytical solution as BC.



## Goals
- To deal with the phase field in the initial conditions to support the `ELBDM_HYBRID` scheme
- To resolve #118 
- To resolve one of the tasks in https://github.com/gamer-project/gamer/issues/299


## Summary of the hybrid scheme in all the ELBDM test problems
- These already supported the hybrid scheme explicitly
   - `LSS`
   - `VortexPairRotating`
   - `VortexPairLinear`
- These initial phases are already unwrapped
   - `Soliton`: zero everywhere
   - `ExtPot`: zero everywhere
   - `SelfSimilarHalo`: unwrapped phase is computed from velocity
   - `Perturbation`: zero everywhere
   - `JeansInstabilityComoving`: small and continuous and periodic
- These unwrapped phases can be found analytically
   - `PlaneWave`
   - `GaussianWavePacket`
- There is no way to unwrap these initial phases
   - `IsolatedHalo`
      - There are vortices in the halo. 
   - `HaloMerger`
      - There are vortices in the halo.
      - The phase at the boundary between the object and empty space is discontinuous.
      - The phase at the boundary between objects is discontinuous.
   - `DiskHeating`
      - There are vortices in the halo.


## Changes
- For the `PlaneWave` test problem
   - Resolve errors related to the passive field "WrappedPhase".
   - Support the analytical unwrapped phase, which is a linear function, for the hybrid scheme.
   - Support user-defined BC with analytical solutions and ensure the traveling wave uses it in the hybrid scheme.
   - Support non-integer number of wavelength.
   - Check there are wave levels in the hybrid scheme to deal with the nodes in the standing wave.
   - Add Gnuplot scripts to plot the wave function and phase.

- For the `GaussianWavePacket` test problem
   - Support analytical solutions of the unwrapped phase of multiple wave packets for the hybrid scheme.
   - Ensure the traveling wave packets use the user-defined BC in the hybrid scheme.
   - Add Gnuplot script to plot the wave function and phase.

- Return error for those test problems that do not support the hybrid scheme, i.e. `IsolatedHalo` and `HaloMerger`.
   - Unless we can refine all the regions with density > 0 to wave level.

- Global phase unwrapping in the initial condition is unnecessary. No test problem needs it currently.



## Issues
- In `GaussianWavePacket`, the fluid scheme may lead to a "time-step too small" error in some cases (e.g. the default setup), probably related to the low-density region.
